File: beakjoon/dp6_longestdown.py
N = int(input())
seq = list(map(int, input().split()))
dp = [0] * N

for i in range(N):
    tmp = 0
    for j in range(i):
        if seq[i] < seq[j]:
            tmp = max(dp[j],tmp)
    dp[i] = tmp + 1
    
print(max(dp))

File: beakjoon/dp6_longestdown.java
import java.util.Scanner;

public class dp6_longestdown {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[1001];
        int[] dp = new int[1001];
        for(int i=1;i<=n;i++){
            arr[i] = sc.nextInt();
        }
        for(int i=1;i<=n;i++){
            int tmp = 0;
            for (int j =1;j<=i;j++){
                if (arr[j] > arr[i]){
                    tmp = Math.max(dp[j],tmp);
                }
            }
            dp[i] = tmp +1;
        }
        int max = 0;
        for(int i =0;i<=n;i++){
            max = Math.max(dp[i],max);
        }
        System.out.println(max);

        sc.close();
    }
    
}
